### ISSUE TITLE
Cannot start Keycloak Container: Build failed due to errors / Method too large

### ISSUE THREAD

**Issue by ReluXingZeng** [keycloak/keycloak#23647]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

core

### Describe the bug

Cannot start Keycloak Container using kc.sh in dev mode.

I got an error saying `Build failed due to errors`
and the build failure says `org.objectweb.asm.MethodTooLargeException: Method too large: io/quarkus/runtime/generated/RunTimeDefaultsConfigSource.<clinit> ()V`

### Version

22.0.1 / 22.0.3 / nightly

### Expected behavior

Running `kc.sh` should start the keycloak server

### Actual behavior

I got the following error when trying to start `kc.sh`:

```
Starting kc.sh...                                                                                                                                     
Updating the configuration and installing your custom providers, if any. Please wait.                                                                 
ERROR: Failed to run 'build' command.                                                                                                                
ERROR: io.quarkus.builder.BuildException: Build failure: Build failed due to errors                                                                           
        [error]: Build step io.quarkus.deployment.steps.ConfigGenerationBuildStep#runtimeDefaultsConfig threw an exception: org.objectweb.asm.MethodTooLargeException: Method too large: io/quarkus/runtime/generated/RunTimeDefaultsConfigSource.<clinit> ()V                                                      
        at org.objectweb.asm.MethodWriter.computeMethodInfoSize(MethodWriter.java:2088)                                                                       
        at org.objectweb.asm.ClassWriter.toByteArray(ClassWriter.java:511)                                                                                    
        at io.quarkus.gizmo.ClassCreator.writeTo(ClassCreator.java:238)
        at io.quarkus.gizmo.ClassCreator.close(ClassCreator.java:249)
        at io.quarkus.deployment.steps.ConfigGenerationBuildStep.generateDefaultsConfigSource(ConfigGenerationBuildStep.java:566)
        at io.quarkus.deployment.steps.ConfigGenerationBuildStep.runtimeDefaultsConfig(ConfigGenerationBuildStep.java:175)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:864)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
        at java.base/java.lang.Thread.run(Thread.java:833)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)

ERROR: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.deployment.steps.ConfigGenerationBuildStep#runtimeDefaultsConfig threw an exception: org.objectweb.asm.MethodTooLargeException: Method too large: io/quarkus/runtime/generated/RunTimeDefaultsConfigSource.<clinit> ()V
        at org.objectweb.asm.MethodWriter.computeMethodInfoSize(MethodWriter.java:2088)
        at org.objectweb.asm.ClassWriter.toByteArray(ClassWriter.java:511)
        at io.quarkus.gizmo.ClassCreator.writeTo(ClassCreator.java:238)
        at io.quarkus.gizmo.ClassCreator.close(ClassCreator.java:249)
        at io.quarkus.deployment.steps.ConfigGenerationBuildStep.generateDefaultsConfigSource(ConfigGenerationBuildStep.java:566)
        at io.quarkus.deployment.steps.ConfigGenerationBuildStep.runtimeDefaultsConfig(ConfigGenerationBuildStep.java:175)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:864)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
        at java.base/java.lang.Thread.run(Thread.java:833)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)

ERROR: Method too large: io/quarkus/runtime/generated/RunTimeDefaultsConfigSource.<clinit> ()V
For more details run the same command passing the '--verbose' option. Also you can use '--help' to see the details about the usage of the particular command.
```

### How to Reproduce?

I run it as a container in my organization's k8s environment in dev mode

Dockerfile I use:
```
FROM registry.access.redhat.com/ubi9 AS ubi-micro-build
RUN mkdir -p /mnt/rootfs
RUN dnf install --installroot /mnt/rootfs curl --releasever 9 --setopt install_weak_deps=false --nodocs -y; dnf --installroot /mnt/rootfs clean all

# Two Layer Build needed due to keycloak image missing vital elements
# https://keycloak.discourse.group/t/keycloak-21-docker-missing-microdnf/20464/2
# https://github.com/ASzc/keycloak/blob/f6d532d25c3dc65d6f180382e01bef3b8ec35bfe/docs/guides/src/main/server/containers.adoc#installing-additional-rpm-packages

FROM quay.io/keycloak/keycloak:nightly

USER root

COPY --from=ubi-micro-build /mnt/rootfs /
RUN echo "keycloak:x:0:root" >> /etc/group && \
    echo "keycloak:x:1000:0:keycloak user:/opt/keycloak:/sbin/nologin" >> /etc/passwd
COPY ./tiq_entrypoint.sh /opt/keycloak/tiq_entrypoint.sh
RUN chmod +x /opt/keycloak/tiq_entrypoint.sh
RUN chown keycloak /opt/keycloak/tiq_entrypoint.sh
COPY ./init_users.sh /opt/keycloak/init_users.sh
RUN chmod +x /opt/keycloak/init_users.sh
RUN chown keycloak /opt/keycloak/init_users.sh
USER keycloak
ENTRYPOINT [ "/opt/keycloak/entrypoint.sh" ]
```

Entrypoint I use:
```
#!/bin/bash
#
# This add more functionality to the build-in keycloak entrypoint, kc.sh
#
set -eou pipefail

# usage: file_env VAR [DEFAULT]
#    ie: file_env 'XYZ_DB_PASSWORD' 'example'
# (will allow for "$XYZ_DB_PASSWORD_FILE" to fill in the value of
#  "$XYZ_DB_PASSWORD" from a file, especially for Docker's secrets feature)
file_env() {
    local var="$1"
    local fileVar="${var}_FILE"
    local def="${2:-}"
    if [[ ${!var:-} && ${!fileVar:-} ]]; then
        echo >&2 "error: both $var and $fileVar are set (but are exclusive)"
        exit 1
    fi
    local val="$def"
    if [[ ${!var:-} ]]; then
        val="${!var}"
    elif [[ ${!fileVar:-} ]]; then
        val="$(< "${!fileVar}")"
    fi
    if [[ -n $val ]]; then
        export "$var"="$val"
    fi
    unset "$fileVar"
}

#############
# Configure #
#############
file_env 'KEYCLOAK_ADMIN'
file_env 'KEYCLOAK_ADMIN_PASSWORD'
# Required, based on https://github.com/keycloak/keycloak/issues/13148#issuecomment-1246864193 and https://keycloak.discourse.group/t/loading-the-admin-console-spinning/17105/6
file_env 'KC_HTTP_RELATIVE_PATH'
file_env 'KC_HOSTNAME_URL'
file_env 'KC_HOSTNAME_ADMIN_URL'

##################
# Start Keycloak #
##################
echo "Starting kc.sh..."
exec /opt/keycloak/bin/kc.sh start-dev --log-level=org.keycloak.events:debug --spi-login-protocol-openid-connect-legacy-logout-redirect-uri=true --proxy=edge --hostname-strict-https=false --http-enabled=true --http-port=8080 --hostname-strict=false
```

k8s deployment specs:
```
  keycloak:                                                                                                                                           Container ID:   [REDACTED]                                                        Image:          [REDACTED]                                                                         Image ID:       [REDACTED]                    Port:           8080/TCP
    Host Port:      0/TCP
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Error
      Exit Code:    1
      Started:      Mon, 02 Oct 2023 19:51:58 -0400
      Finished:     Mon, 02 Oct 2023 19:52:15 -0400
    Ready:          False
    Restart Count:  6
    Limits:
      cpu:     1
      memory:  1Gi
    Requests:
      cpu:     1
      memory:  1Gi
    Environment:
      KC_HTTP_RELATIVE_PATH:         /keycloak/auth/
      KC_HOSTNAME_URL:               https://[REDACTED]
      KC_HOSTNAME_ADMIN_URL:         https://[REDACTED]
      KEYCLOAK_ADMIN_PASSWORD_FILE:  /run/secrets/loginadmin/loginadmin
      KEYCLOAK_ADMIN:                admin
    Mounts:
      /run/secrets/loginadmin from loginadmin-[REDACTED] (ro)
```

### Anything else?

I am not the only person seeing this issue, while I was searching online I found:
https://groups.google.com/g/keycloak-user/c/FyxLLKSNRq4
We have exactly the same issue

Also, worth noting that, our keycloak container **was** working, at least 4 days ago - in September 28th. Something happened during the 4 days cause it to broke. And what's more interesting about it is that, we didn't rebuild the image within the 4 days, so the the image tag didn't change. This means that the image that's working in September 28th is the same as the image that's breaking on October 2nd. Not sure if it means that the data it was trying to fetch here is from a remote source. I don't think it's from our source though, as the KC_HOSTNAME_URL and KC_HOSTNAME_ADMIN_URL are the only two url we provided, and they all point to keycloak itself.

**Comment by ReluXingZeng** [comment]
This issue seems to be gone once we downgrade to 21.x

We will downgrade for now, but hopefully this can be fixed for 22.x

**Comment by vmuzikar** [comment]
@shawkins Could you please look into this when you get a chance?

**Comment by shawkins** [comment]
Something similar has been observed with quarkus before https://github.com/quarkusio/quarkus/issues/22027

Running a quick test locally I see quarkus trying to create a constructor that is handling 133 runtime defaults.  That exception is possible for @ReluXingZeng if somehow a much larger number of values are being picked up.

Looking at the quarkus code, the configItem.getReadResult().getRunTimeDefaultValues() seems to use the entire env at the time it is run + additional entries.  Perhaps do something like "echo `env | wc -l`" in your entry point to see if you are picking up an unsually large number of env variables.

**Comment by ReluXingZeng** [comment]
I see, so you are suggesting that it could be resulted from there being a lot of environmental variables that are used in our k8s environment might result in the sudden failure of Keycloak. 

There does seems to be a lot of environmental variables in our k8s cluster....and indeed more than what we had before we have the failure. 

```
bash-5.1$ env | wc -l
5166
```

Any suggestion of how much "space" a environmental variable make take? I recall the limitation here in Java is a maximum of 64KB for a single method. If every environmental variable is just somehow placed inside the constructor, then 5166 x probably al least more than 13 bytes on average could easily give 67KB.

**Comment by shawkins** [comment]
> If every environmental variable is just somehow placed inside the constructor, then 5166 x probably al least more than 13 bytes on average could easily give 67KB.

Yes, over 5000 env variables could easily cause the method to exceed the java maximum.  Do you have a lot of Services? https://kubernetes.io/docs/concepts/services-networking/service/#environment-variables 

If it seems like you'll need support for this amount of env variables, this issue will need to be moved to quarkus for further work.  For example they can create a method for every n defaults and call that in the constructor.

**Comment by ReluXingZeng** [comment]
Make sense! I have a very limited knowledge on what quarkus is, I will see how I should raise a ticket there.

In the meantime, I am thinking if I can use the following code to reduce the environmental variable flowing from k8s:

```echo "Number of environment variable before deletion: $(env | wc -l)"
env_list=($(env | grep "SOMETHING_FOR_REMOVAL"))
for var_value in "${env_list[@]}"
do
    var=${var_value%%=*}
    unset $var
done
echo "Number of environment variable after deletion: $(env | wc -l)"```
I _think_, but not entirely confident, this will make JVM pickup less variable to prevent qurakus casuing issue on this.

Somehow, the number of bash variable we have in our dev environment these days has fall below 4000 after we removed some services. Seems like this is the number that in either way we can boost our keycloak services without issue.

**Comment by shawkins** [comment]
Closing on the issue on the keycloak side.  When you open the quarkus issue be sure to metion this one. 

cc @geoand

**Comment by vmuzikar** [comment]
Keycloak sticks to Quarkus LTS versions. That means we'll be on Quarkus 3.2 up to Keycloak 24 where we'll upgrade to the next LTS.

**Comment by saurabhPatel1712** [comment]
Hey @vmuzikar, @shawkins @ReluXingZeng, I'm facing a similar issue while using Keycloak 22.0.5. Would you like me to upgrade to the later version, or do we have some solution identified ? Need your support and guidance.

We're running Keycloak in a Kubernetes cluster using a custom Docker setup. Our Dockerfile downloads the Keycloak JAR and executes the following build and start commands:

/keycloak/bin/kc.sh build --http-relative-path=/auth --health-enabled=true --metrics-enabled=true

/keycloak/bin/kc.sh start --hostname-strict=false --https-port=8000 --hostname-port=8000  --hostname-admin-url=some-admin-url --hostname-url=some-url --proxy=edge --http-enabled=true --https-certificate-file=service.crt --https-certificate-key-file=service.key --import-realm --log="console,file"  &

`
2025-02-06 15:18:35,503 [main] WARN  co.elastic.apm.agent.configuration.StartupInfo - To enable all features and decrease startup time, please configure application_packages
2025-02-06 15:18:36,505 [elastic-apm-server-healthcheck] INFO  co.elastic.apm.agent.report.ApmServerHealthChecker - Elastic APM server is available: {  "build_date": "2022-01-06T23:27:35Z",  "build_sha": "fde0af4fa2b9f39e518b333c5be56cf8be215ca0",  "publish_ready": true,  "version": "7.16.3"}
2025-02-06 15:18:36,523 [main] INFO  co.elastic.apm.agent.impl.ElasticApmTracer - Tracer switched to RUNNING state
Updating the configuration and installing your custom providers, if any. Please wait.
ERROR: Failed to run 'build' command.
ERROR: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.deployment.steps.ConfigGenerationBuildStep#runtimeDefaultsConfig threw an exception: org.objectweb.asm.MethodTooLargeException: Method too large: io/quarkus/runtime/generated/RunTimeDefaultsConfigSource.<clinit> ()V
	at org.objectweb.asm.MethodWriter.computeMethodInfoSize(MethodWriter.java:2088)
	at org.objectweb.asm.ClassWriter.toByteArray(ClassWriter.java:511)
	at io.quarkus.gizmo.ClassCreator.writeTo(ClassCreator.java:238)
	at io.quarkus.gizmo.ClassCreator.close(ClassCreator.java:249)
	at io.quarkus.deployment.steps.ConfigGenerationBuildStep.generateDefaultsConfigSource(ConfigGenerationBuildStep.java:566)
	at io.quarkus.deployment.steps.ConfigGenerationBuildStep.runtimeDefaultsConfig(ConfigGenerationBuildStep.java:175)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:864)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at java.base/java.lang.Thread.run(Thread.java:840)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

ERROR: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.deployment.steps.ConfigGenerationBuildStep#runtimeDefaultsConfig threw an exception: org.objectweb.asm.MethodTooLargeException: Method too large: io/quarkus/runtime/generated/RunTimeDefaultsConfigSource.<clinit> ()V
	at org.objectweb.asm.MethodWriter.computeMethodInfoSize(MethodWriter.java:2088)
	at org.objectweb.asm.ClassWriter.toByteArray(ClassWriter.java:511)
	at io.quarkus.gizmo.ClassCreator.writeTo(ClassCreator.java:238)
	at io.quarkus.gizmo.ClassCreator.close(ClassCreator.java:249)
	at io.quarkus.deployment.steps.ConfigGenerationBuildStep.generateDefaultsConfigSource(ConfigGenerationBuildStep.java:566)
	at io.quarkus.deployment.steps.ConfigGenerationBuildStep.runtimeDefaultsConfig(ConfigGenerationBuildStep.java:175)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:864)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at java.base/java.lang.Thread.run(Thread.java:840)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)

ERROR: Method too large: io/quarkus/runtime/generated/RunTimeDefaultsConfigSource.<clinit> ()V
For more details run the same command passing the '--verbose' option. Also you can use '--help' to see the details about the usage of the particular command.
`

**Comment by shawkins** [comment]
> Would you like me to upgrade to the later version

Yes, you should upgrade. This was addressed by later Quarkus changes.

**Comment by shawkins** [comment]
See the previous comments about reducing the number of environment variables https://github.com/keycloak/keycloak/issues/23647#issuecomment-1754238446

**Comment by saurabhPatel1712** [comment]
Thank you, @shawkins. Your input helped me obtain the environment details of the Keycloak Kubernetes pods. I've noticed that our Keycloak pods have a total of 5,675 environment variables. Some are specific to Keycloak, while the rest are default Kubernetes environment variables in the particular namespace.

Is this issue resolved in Keycloak version 23.0.0? Which version would you recommend I try: 23.0.0 or 23.0.7?

**Comment by shawkins** [comment]
> Is this issue resolved in Keycloak version 23.0.0? Which version would you recommend ?

The fix was not picked up until Keycloak 24 - https://github.com/keycloak/keycloak/issues/23647#issuecomment-1780548365

We always recommend using the latest release possible.
